internal/cloud/s3.go

```diff
@@ -0,0 +1,401 @@
+		if customName == path {
+			return fmt.Errorf("unexpected path: `%v` triming `%v`", path, localDirectoy)
+		}
+		customName = filepath.ToSlash(customName)
+		if len(customName) == 0 || customName[0] != '/' {
+			customName = "/" + customName
+		}
+		customName = base + customName
+
+		if info.IsDir() {
+			return compression.Write(archiver.File{
+				FileInfo: archiver.FileInfo{
+					FileInfo:   info,
+					CustomName: customName,
+				},
+			})
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		return compression.Write(archiver.File{
+			FileInfo: archiver.FileInfo{
+				FileInfo:   info,
+				CustomName: customName,
+			},
+			ReadCloser: f,
+		})
+	})
+	if err != nil {
+		return err
+	}
+
+	compression.Close()
+	if closer != nil {
+		closer.Close()
+	}
+	wg.Wait()
+
+	uploader := s3manager.NewUploader(a.s)
+
+	a.m.Lock()
+	ctxt, cancel := context.WithCancel(context.Background())
+	a.cancel = cancel
+	a.m.Unlock()
+	defer a.Cancel()
+
+	f, err := os.Open(file.Name())
+	if err != nil {
+		return err
+	}
+
+	_, err = uploader.UploadWithContext(ctxt, &s3manager.UploadInput{
+		Bucket: aws.String(a.bucket),
+		Key:    aws.String(s3FilePath),
+
+		Body: f,
+	})
+	if err != nil {
+		return err
+	}
+	f.Close()
+
+	return err
+}
+
+func (a *AWSSession) DownloadFile(s3FilePath string, localFile string) error {
+	f, err := os.Create(localFile)
+	if err != nil {
+		return err
+	}
+
+	downloader := s3manager.NewDownloader(a.s)
+
+	a.m.Lock()
+	ctxt, cancel := context.WithCancel(context.Background())
+	a.cancel = cancel
+	a.m.Unlock()
+	defer a.Cancel()
+
+	_, err = downloader.DownloadWithContext(ctxt, f, &s3.GetObjectInput{
+		Bucket: aws.String(a.bucket),
+		Key:    aws.String(s3FilePath),
+	})
+
+	return err
+}
+
+func (a *AWSSession) DownloadCompressedDirectory(s3FilePath string, localRootDirectory string) error {
+	file, err := ioutil.TempFile("", "fyne-cross-s3")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(file.Name())
+
+	a.m.Lock()
+	ctxt, cancel := context.WithCancel(context.Background())
+	a.cancel = cancel
+	a.m.Unlock()
+	defer a.Cancel()
+
+	downloader := s3manager.NewDownloader(a.s)
+	downloader.Concurrency = 1
+
+	_, err = downloader.DownloadWithContext(ctxt, fakeWriterAt{file}, &s3.GetObjectInput{
+		Bucket: aws.String(a.bucket),
+		Key:    aws.String(s3FilePath),
+	})
+	file.Close()
+	if err != nil {
+		return err
+	}
+
+	in, err := os.Open(file.Name())
+	if err != nil {
+		return err
+	}
+
+	extension := strings.ToLower(filepath.Ext(s3FilePath))
+
+	var compression archiver.Reader
+
+	switch extension {
+	case ".xz":
+		compression = archiver.NewTarXz()
+		err := compression.Open(in, 0)
+		if err != nil {
+			return err
+		}
+	case ".zstd":
+		inTar, outZstd := io.Pipe()
+
+		dec, err := zstd.NewReader(in)
+		if err != nil {
+			return err
+		}
+		defer dec.Close()
+
+		go func() {
+			// Copy content...
+			io.Copy(outZstd, dec)
```

Comment:
I think this is an ignored error...

Reply:
Good catch, using an error group solve this.